indico/modules/rb_new/controllers/backend/blockings.py

```diff
@@ -0,0 +1,130 @@
+from indico.modules.rb.notifications.blockings import notify_request
+from indico.modules.rb_new.operations.blockings import get_room_blockings
+from indico.modules.rb_new.schemas import blockings_schema
+from indico.modules.users.models.users import User
+from indico.web.util import jsonify_data
+
+
+def _group_id_or_name(principal):
+    provider = principal['provider']
+    if provider is None or provider == 'indico':
+        return principal['id']
+    else:
+        return principal['name']
+
+
+@no_autoflush
+def _populate_blocking(blocking, room_ids, allowed_principals, reason):
+    blocking.blocked_rooms = [BlockedRoom(room_id=room.id) for room in Room.query.filter(Room.id.in_(room_ids))]
+    blocking.reason = reason
+    blocking.allowed = [GroupProxy(_group_id_or_name(pr), provider=pr['provider'])
+                        if pr.get('is_group')
+                        else User.get_one(pr['id'])
+                        for pr in allowed_principals]
+
+
+def create_blocking(room_ids, allowed_principals, start_date, end_date, reason, created_by):
+    blocking = Blocking()
+    blocking.start_date = start_date
+    blocking.end_date = end_date
+    blocking.created_by_user = created_by
+    _populate_blocking(blocking, room_ids, allowed_principals, reason)
+    db.session.add(blocking)
+    db.session.flush()
+    return blocking
+
+
+def update_blocking(blocking, room_ids, allowed_principals, reason):
+    _populate_blocking(blocking, room_ids, allowed_principals, reason)
+    db.session.flush()
+
+
+def approve_or_request_blocking(blocking):
+    rooms_by_owner = defaultdict(list)
+    for blocked_room in blocking.blocked_rooms:
+        owner = blocked_room.room.owner
+        if owner == session.user:
+            blocked_room.approve(notify_blocker=False)
+        else:
+            rooms_by_owner[owner].append(blocked_room)
+
+    for owner, rooms in rooms_by_owner.iteritems():
+        notify_request(owner, blocking, rooms)
+
+
+class RHCreateRoomBlocking(RHRoomBookingBase):
+    @use_args({
+        'room_ids': fields.List(fields.Int(), missing=[]),
+        'start_date': fields.Date(),
+        'end_date': fields.Date(),
+        'reason': fields.Str(),
+        'allowed_principals': fields.List(fields.Dict())
+    })
+    def _process(self, args):
+        blocking = create_blocking(created_by=session.user, **args)
+        approve_or_request_blocking(blocking)
+        return jsonify_data(flash=False)
+
+
+class RHUpdateRoomBlocking(RHRoomBookingBase):
+    def _check_access(self):
+        RHRoomBookingBase._check_access(self)
+        if not self.blocking.can_be_modified(session.user):
+            raise Forbidden
+
+    def _process_args(self):
+        self.blocking = Blocking.get_one(request.view_args['blocking_id'])
+
+    @use_args({
+        'room_ids': fields.List(fields.Int(), required=True),
+        'reason': fields.Str(required=True),
+        'allowed_principals': fields.List(fields.Dict(), missing=[])
```

Comment:
Unrelated to this PR: Can't (and shouldn't) we specify a proper schema for the list entries? Maybe something like `fields.Nested`?

Reply:
You meant the equivalent of PropTypes.shape({...})? Yeah, that would be nice...